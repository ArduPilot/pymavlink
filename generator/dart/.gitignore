# Files and directories created by pub.
.dart_tool/
.packages
doc/api/
coverage/

# Conventional directory for build outputs.
build/

# Omit committing pubspec.lock for library packages; see
# https://dart.dev/guides/libraries/private-files#pubspeclock.
pubspec.lock
