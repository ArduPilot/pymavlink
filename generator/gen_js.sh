#!/bin/sh

for protocol in 1.0 2.0; do
 for xml in ../../message_definitions/v1.0/*.xml; do
     base=$(basename $xml .xml)
     mkdir -p javascript/implementations/mavlink_${base}_v${protocol}

     # Generate MAVLink implementation
     ../tools/mavgen.py --lang=JavaScript_NextGen --wire-protocol=$protocol --output=javascript/implementations/mavlink_${base}_v${protocol}/mavlink.js $xml || exit 1



     # Create package.json file
     cat >javascript/implementations/mavlink_${base}_v${protocol}/package.json <<EOF
 {
    "name" : "mavlink_${base}_v${protocol}",
    "version" : "0.0.3",
    "description" : "NextGen Implementation of the MAVLink protocol",
    "keywords" : ["mavlink", "arduino", "ardupilot", "ros", "robot", "uav", "drone", "awesome"],
    "homepage": "https://github.com/ardupilot/mavlink",
    "bugs" : "https://github.com/ardupilot/mavlink/issues",
    "license" : {
        "type" : "LGPL-3.0",
        "url" : "http://opensource.org/licenses/LGPL-3.0"
      },
    "contributors" : ["Bruce Crevensten <bruce.crevensten@gmail.com>","David Buzz <davidbuzz@gmail.com>"],
    "main" : "mavlink.js",
    "repository" : {
      "type" : "git",
      "url" : "https://github.com/ardupilot/mavlink"
      },
    "dependencies" : {
      "underscore" : "",
      "winston": "",
      "jspack": "file:local_modules/jspack",
      "long": "file:local_modules/long"
    },
    "devDependencies" : {
      "should" : "",
      "mocha" : "",
      "sinon" : ""
    }
}
EOF

    #symlink our custom jspack and 'long' so the package.json above can find them without needing multiple copies
    rm ./javascript/implementations/mavlink_${base}_v${protocol}/local_modules
    ln -s ../../local_modules ./javascript/implementations/mavlink_${base}_v${protocol}/

 done
done

