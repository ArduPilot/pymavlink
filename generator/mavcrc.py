'''
MAVLink CRC-16/MCRF4XX code

Copyright Andrew Tridgell
Released under GNU LGPL version 3 or later
'''
import sys
from builtins import object


class x25crc(object):
    """CRC-16/MCRF4XX - based on checksum.h from mavlink library"""

    def __init__(self, buf=None):
        self.crc = 0xFFFF
        if buf is not None:
            self.accumulate(buf)

    def accumulate(self, buf):
        """add in some more bytes (it also accepts strings)"""
        if sys.version_info[0] == 2:
            if type(buf) is str:
                buf = bytearray(buf)
            elif type(buf).__name__ == 'unicode':
                # we can't use the identifier unicode in python3
                buf = bytearray(buf.encode())
        elif type(buf) is str:
            buf = buf.encode()

        accum = self.crc
        for b in buf:
            tmp = b ^ (accum & 0xFF)
            tmp = (tmp ^ (tmp << 4)) & 0xFF
            accum = (accum >> 8) ^ (tmp << 8) ^ (tmp << 3) ^ (tmp >> 4)
        self.crc = accum

    def accumulate_str(self, buf):
        """
        Provided for backwards compatibility. accumulate now also works on strings.
        """
        return self.accumulate(buf)
