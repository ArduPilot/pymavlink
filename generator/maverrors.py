#!/usr/bin/env python
'''
Define MAVError and MAVParseError
Copyright Andrew Tridgell 2011
Released under GNU GPL version 3 or later
'''


class MAVError(Exception):
        '''MAVLink error class'''
        def __init__(self, msg):
            Exception.__init__(self, msg)
            self.message = msg


class MAVParseError(Exception):
    def __init__(self, message, inner_exception=None):
        self.message = message
        self.inner_exception = inner_exception
        self.exception_info = sys.exc_info()
    def __str__(self):
        return self.message
