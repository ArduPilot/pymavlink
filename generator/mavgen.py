#!/usr/bin/env python

'''parse a MAVLink protocol XML file and generate a python implementation

Copyright Andrew Tridgell 2011
Released under GNU GPL version 3 or later

General process:
 - each filename passed in:
    - may be validated, based on --validate
    - is parsed using mavparse.MAVXML into an xml document and appended to a list, "xml"

 - expand_includes is called to do a breadth-first search of the xml
    documents structure inferred by the <include> tags in each
    document, expanding the xml list from its base (just the ones on
    the commandline) to the entire structure

 - update_includes works on the xml list created by expand_includes
    - any xml document with no includes is added to the "done" list (there must be at least one of these)
    - it repeatedly calls update_one_iteration
    - each iteration is intended to include the crcs and other information from includes into the xml document doing the include

'''

from __future__ import print_function
from future import standard_library
standard_library.install_aliases()
from builtins import object
import os
import re
import sys
from . import mavparse

# XSD schema file
schemaFile = os.path.join(os.path.dirname(os.path.realpath(__file__)), "mavschema.xsd")

# Set defaults for generating MAVLink code
DEFAULT_WIRE_PROTOCOL = mavparse.PROTOCOL_1_0
DEFAULT_LANGUAGE = 'Python'
DEFAULT_ERROR_LIMIT = 200
DEFAULT_VALIDATE = True
DEFAULT_STRICT_UNITS = False

MAXIMUM_INCLUDE_FILE_NESTING = 5

# List the supported languages. This is done globally because it's used by the GUI wrapper too
supportedLanguages = ["C", "CS", "JavaScript", "TypeScript", "Python", "Lua", "WLua", "ObjC", "Swift", "Java", "C++11"]


def mavgen(opts, args):
    """Generate mavlink message formatters and parsers (C and Python ) using options
    and args where args are a list of xml files. This function allows python
    scripts under Windows to control mavgen using the same interface as
    shell scripts under Unix"""

    xml = []
    all_files = set()

    # Enable validation by default, disabling it if explicitly requested
    if opts.validate:
        try:
            from lxml import etree
            with open(schemaFile, 'r') as f:
                xmlschema_root = etree.parse(f)
                if not opts.strict_units:
                    # replace the strict "SI_Unit" list of known unit strings with a more generic "xs:string" type
                    for elem in xmlschema_root.iterfind('xs:attribute[@name="units"]', xmlschema_root.getroot().nsmap):
                        elem.set("type", "xs:string")
                xmlschema = etree.XMLSchema(xmlschema_root)
        except ImportError:
            print("WARNING: Failed to import lxml module etree. Are lxml, libxml2 and libxslt installed? XML validation will not be performed", file=sys.stderr)
            opts.validate = False
        except etree.XMLSyntaxError as err:
            print("WARNING: XML Syntax Errors detected in %s XML schema file. XML validation will not be performed" % schemaFile, file=sys.stderr)
            print(str(err.error_log), file=sys.stderr)
            opts.validate = False
        except Exception as e:
            print("Exception:", e)
            print("WARNING: Unable to load XML validator libraries. XML validation will not be performed", file=sys.stderr)
            opts.validate = False

    def expand_includes():
        """Expand includes. Root files already parsed objects in the xml list."""

        def expand_oneiteration():
            '''takes the list of xml files to process and finds includes which
            have not already been turned into xml documents added to
            xml files to process, turns them into xml documents and
            adds them to the xml files list.  Returns false if no more
            documents were added.
            '''
            includeadded = False
            for x in xml[:]:
                for i in x.include:
                    fname = os.path.join(os.path.dirname(x.filename), i)
                    # Only parse new include files
                    if fname in all_files:
                        continue
                    # Validate XML file with XSD file if possible.
                    if opts.validate:
                        print("Validating %s" % fname)
                        if not mavgen_validate(fname):
                            print("ERROR Validation of %s failed" % fname)
                            exit(1)
                    else:
                        print("Validation skipped for %s." % fname)
                    # Parsing
                    print("Parsing %s" % fname)
                    xml.append(mavparse.MAVXML(fname, opts.wire_protocol))
                    all_files.add(fname)
                    includeadded = True
            return includeadded

        for i in range(MAXIMUM_INCLUDE_FILE_NESTING):
            if not expand_oneiteration():
                break

        if mavparse.check_duplicates(xml):
            sys.exit(1)

    def update_includes():
        """Update dialects with crcs etc of included files.  Included files
        were already found and parsed into xml list in
        expand_includes().
        """

        # 1: Mark files that don't have includes as "done"
        done = []
        for x in xml:
            #print("\n",x)
            if len(x.include) == 0:
                done.append(x)
                #print("\nFile with no includes found (ENDPOINT): %s" % x.filename )
        if len(done) == 0:
            print("\nERROR in includes tree, no base found!")
            exit(1)

        #print("\n",done)

        # 2: Update all 'not done' files for which all includes have
        # been done.  Returns True if any updates were made
        def update_oneiteration():
            initial_done_length = len(done)
            for x in xml:
                #print("\nCHECK %s" % x.filename)
                if x in done:
                    #print("  already done, skip")
                    continue
                #check if all its includes were already done
                all_includes_done = True
                for i in x.include:
                    fname = os.path.join(os.path.dirname(x.filename), i)
                    if fname not in [d.filename for d in done]:
                        all_includes_done = False
                        break
                if not all_includes_done:
                    #print("  not all includes ready, skip")
                    continue
                #Found file where all includes are done
                done.append(x)
                #print("  all includes ready, add" )
                #now update it with the facts from all it's includes
                for i in x.include:
                    fname = os.path.join(os.path.dirname(x.filename), i)
                    #print("  include file %s" % i )
                    #Find the corresponding x
                    for ix in xml:
                        if ix.filename != fname:
                            continue
                        #print("    add %s" % ix.filename )
                        x.message_crcs.update(ix.message_crcs)
                        x.message_lengths.update(ix.message_lengths)
                        x.message_min_lengths.update(ix.message_min_lengths)
                        x.message_flags.update(ix.message_flags)
                        x.message_target_system_ofs.update(ix.message_target_system_ofs)
                        x.message_target_component_ofs.update(ix.message_target_component_ofs)
                        x.message_names.update(ix.message_names)
                        x.largest_payload = max(x.largest_payload, ix.largest_payload)
                        break

            if len(done) == len(xml):
                return False  # finished
            if len(done) == initial_done_length:
                # we've made no progress
                print("ERROR include tree can't be resolved, no base found!")
                exit(1)
            return True

        for i in range(MAXIMUM_INCLUDE_FILE_NESTING):
            #print("\nITERATION "+str(i))
            if not update_oneiteration():
                break

    def mavgen_validate(xmlfile):
        """Uses lxml to validate an XML file. We define mavgen_validate
           here because it relies on the XML libs that were loaded in mavgen(), so it can't be called standalone"""
        xmlvalid = True
        try:
            with open(xmlfile, 'r') as f:
                xmldocument = etree.parse(f)
                xmlschema.assertValid(xmldocument)
                forbidden_names_re = re.compile("^(break$|case$|class$|catch$|const$|continue$|debugger$|default$|delete$|do$|else$|\
                                    export$|extends$|finally$|for$|function$|if$|import$|in$|instanceof$|let$|new$|\
                                    return$|super$|switch$|this$|throw$|try$|typeof$|var$|void$|while$|with$|yield$|\
                                    enum$|await$|implements$|package$|protected$|static$|interface$|private$|public$|\
                                    abstract$|boolean$|byte$|char$|double$|final$|float$|goto$|int$|long$|native$|\
                                    short$|synchronized$|transient$|volatile$).*", re.IGNORECASE)
                for element in xmldocument.iter('enum', 'entry', 'message', 'field'):
                    if forbidden_names_re.search(element.get('name')):
                        print("Validation error:", file=sys.stderr)
                        print("Element : %s at line : %s contains forbidden word" % (element.tag, element.sourceline), file=sys.stderr)
                        xmlvalid = False

            return xmlvalid
        except etree.XMLSchemaError:
            return False
        except etree.DocumentInvalid as err:
            sys.exit('ERROR: %s' % str(err.error_log))
        return True

    # Process all XML files, validating them as necessary.
    for fname in args:
        # only add each dialect file argument once.
        if fname in all_files:
            continue
        all_files.add(fname)

        if opts.validate:
            print("Validating %s" % fname)
            if not mavgen_validate(fname):
                return False
        else:
            print("Validation skipped for %s." % fname)

        print("Parsing %s" % fname)
        xml.append(mavparse.MAVXML(fname, opts.wire_protocol))

    # expand includes
    expand_includes()
    update_includes()

    print("Found %u MAVLink message types in %u XML files" % (
        mavparse.total_msgs(xml), len(xml)))

    # convert language option to lowercase and validate
    opts.language = opts.language.lower()
    if opts.language == 'python':
        from . import mavgen_python
        mavgen_python.generate(opts.output, xml)
    elif opts.language == 'c':
        from . import mavgen_c
        mavgen_c.generate(opts.output, xml)
    elif opts.language == 'lua':
        from . import mavgen_lua
        mavgen_lua.generate(opts.output, xml)
    elif opts.language == 'wlua':
        from . import mavgen_wlua
        mavgen_wlua.generate(opts.output, xml)
    elif opts.language == 'cs':
        from . import mavgen_cs
        mavgen_cs.generate(opts.output, xml)
    elif opts.language == 'javascript':
        from . import mavgen_javascript
        mavgen_javascript.generate(opts.output, xml)
    elif opts.language == 'typescript':
        from . import mavgen_typescript
        mavgen_typescript.generate(opts.output, xml)
    elif opts.language == 'objc':
        from . import mavgen_objc
        mavgen_objc.generate(opts.output, xml)
    elif opts.language == 'swift':
        from . import mavgen_swift
        mavgen_swift.generate(opts.output, xml)
    elif opts.language == 'java':
        from . import mavgen_java
        mavgen_java.generate(opts.output, xml)
    elif opts.language == 'c++11':
        from . import mavgen_cpp11
        mavgen_cpp11.generate(opts.output, xml)
    else:
        print("Unsupported language %s" % opts.language)

    return True


# build all the dialects in the dialects subpackage
class Opts(object):
    def __init__(self, output, wire_protocol=DEFAULT_WIRE_PROTOCOL, language=DEFAULT_LANGUAGE, validate=DEFAULT_VALIDATE, error_limit=DEFAULT_ERROR_LIMIT, strict_units=DEFAULT_STRICT_UNITS):
        self.wire_protocol = wire_protocol
        self.error_limit = error_limit
        self.language = language
        self.output = output
        self.validate = validate
        self.strict_units = strict_units


def mavgen_python_dialect(dialect, wire_protocol):
    '''generate the python code on the fly for a MAVLink dialect'''
    dialects = os.path.join(os.path.dirname(os.path.realpath(__file__)), '..', 'dialects')
    mdef = os.path.join(os.path.dirname(os.path.realpath(__file__)), '..', '..', 'message_definitions')
    if wire_protocol == mavparse.PROTOCOL_0_9:
        py = os.path.join(dialects, 'v09', dialect + '.py')
        xml = os.path.join(dialects, 'v09', dialect + '.xml')
        if not os.path.exists(xml):
            xml = os.path.join(mdef, 'v0.9', dialect + '.xml')
    elif wire_protocol == mavparse.PROTOCOL_1_0:
        py = os.path.join(dialects, 'v10', dialect + '.py')
        xml = os.path.join(dialects, 'v10', dialect + '.xml')
        if not os.path.exists(xml):
            xml = os.path.join(mdef, 'v1.0', dialect + '.xml')
    else:
        py = os.path.join(dialects, 'v20', dialect + '.py')
        xml = os.path.join(dialects, 'v20', dialect + '.xml')
        if not os.path.exists(xml):
            xml = os.path.join(mdef, 'v1.0', dialect + '.xml')
    opts = Opts(py, wire_protocol)

    # Python 2 to 3 compatibility
    try:
        import StringIO as io
    except ImportError:
        import io

    # throw away stdout while generating
    stdout_saved = sys.stdout
    sys.stdout = io.StringIO()
    try:
        xml = os.path.relpath(xml)
        if not mavgen(opts, [xml]):
            sys.stdout = stdout_saved
            return False
    except Exception:
        sys.stdout = stdout_saved
        raise
    sys.stdout = stdout_saved
    return True

if __name__ == "__main__":
    raise DeprecationWarning("Executable was moved to pymavlink.tools.mavgen")
