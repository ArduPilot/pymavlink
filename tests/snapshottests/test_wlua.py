#!/usr/bin/env python

"""
Test that the wlua generator works in Wireshark
"""
import os
import shutil
import subprocess

import pytest
from pymavlink.generator import mavgen


def generate_wlua(xml_filepath, dest_dir):
    """
    Create the dissector script
    """
    lua_filepath = dest_dir / "mavlink.lua"
    assert not lua_filepath.is_file()
    mavgen.mavgen(
        mavgen.Opts(output=str(lua_filepath), language="wlua"), [xml_filepath]
    )
    assert lua_filepath.is_file()
    return lua_filepath


try:
    from syrupy import snapshot
except ImportError:

    @pytest.fixture
    def snapshot(*args, **kwargs):
        pytest.skip("syrupy snapshot plugin not available")


@pytest.mark.parametrize(
    ("mdef", "pcap"),
    [
        # simple baseline tests
        ("common.xml", "common.pcapng"),
        # tests both command_long and that we can interpret tcp packets
        ("common.xml", "command_long_over_tcp.pcapng"),
        # test that output is okay when we encounter messages not defined in the xml
        ("minimal.xml", "common.pcapng"),
    ],
)
def test_wlua(request, tmp_path, snapshot, mdef, pcap):
    files = request.path.parent / "resources"
    lua_filepath = generate_wlua(files / mdef, tmp_path)

    packet_file = files / pcap
    no_such_dir = tmp_path / "empty"
    tshark = shutil.which("tshark")
    if tshark is None:
        pytest.skip("tshark (Wireshark CLI) not found")

    env = {
        # suppress system-wide wireshark config (if any)
        "WIRESHARK_CONFIG_DIR": str(no_such_dir),
        # suppress user wireshark config on macOS (and I think Linux)
        "HOME": str(no_such_dir),
    }
    try:
        # needed on Windows to initialize GnuTLS
        env["SYSTEMROOT"] = os.environ["SYSTEMROOT"]
    except KeyError:
        pass

    actual = subprocess.run(
        [
            tshark,
            "-Ttext",
            "-r" + str(packet_file),
            "-Xlua_script:" + str(lua_filepath),
            "-Omavlink_proto",
            "-n",
        ],
        env=env,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        universal_newlines=True,
        check=False,
    )
    # note that, with text output, tshark truncates hex dump lines at 80-ish or 100-ish characters.
    # Truncate them preemptively so this isn't reflected in the diff.
    # This doesn't lose any info we really care about.
    truncated_stdout = ''.join([line[:72]+"\n" for line in actual.stdout.splitlines()])
    
    props_to_match = {
        "stdout": truncated_stdout,
        "stderr": actual.stderr,
        "returncode": actual.returncode,
    }

    assert props_to_match == snapshot()


if __name__ == "__main__":
    pytest.main([__file__])
