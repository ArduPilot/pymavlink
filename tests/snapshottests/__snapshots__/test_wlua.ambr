# serializer version: 1
# name: test_wlua[common.xml-command_long_over_tcp.pcapng]
  dict({
    'returncode': 0,
    'stderr': '',
    'stdout': '''
      Frame 1: 101 bytes on wire (808 bits), 101 bytes captured (808 bits) on 
      Null/Loopback
      Internet Protocol Version 4, Src: 127.0.0.1, Dst: 127.0.0.1
      Transmission Control Protocol, Src Port: 62025, Dst Port: 5760, Seq: 1, 
      MAVLink Protocol (45)
          Header
              Magic value / version: MAVLink 2.0 (0xfd)
              Payload length: 33
              Incompatibility flag: 0
              Compatibility flag: 0
              Packet sequence: 246
              System id: 0xfa
              Component id: 0xfa
              Message id: COMMAND_LONG (76)
          Payload: COMMAND_LONG (76)
              target_system (uint8_t): 1
              target_component (uint8_t): 1
              command (MAV_CMD): MAV_CMD_DO_SET_MODE (176)
              confirmation (uint8_t): 1
              param1: Mode (MAV_MODE): Unknown (1)
              param2: Custom Mode (float): 5
              param3: Custom Submode (float): 0
              param4 (float): 0
              param5 (float): 0
              param6 (float): 0
              param7 (float): 0
          Message CRC: 0xdb93
      
      Frame 2: 138 bytes on wire (1104 bits), 138 bytes captured (1104 bits) o
      Null/Loopback
      Internet Protocol Version 4, Src: 127.0.0.1, Dst: 127.0.0.1
      Transmission Control Protocol, Src Port: 5760, Dst Port: 62025, Seq: 1, 
      MAVLink Protocol (82)
          Header
              Magic value / version: MAVLink 2.0 (0xfd)
              Payload length: 10
              Incompatibility flag: 0
              Compatibility flag: 0
              Packet sequence: 218
              System id: 0x01
              Component id: 0x01
              Message id: COMMAND_ACK (77)
          Payload: COMMAND_ACK (77)
              command (MAV_CMD): MAV_CMD_DO_SET_MODE (176)
              result (MAV_RESULT): MAV_RESULT_FAILED (4)
          Message CRC: 0x7692
      MAVLink Protocol (82)
          Header
              Magic value / version: MAVLink 2.0 (0xfd)
              Payload length: 48
              Incompatibility flag: 0
              Compatibility flag: 0
              Packet sequence: 219
              System id: 0x01
              Component id: 0x01
              Message id: STATUSTEXT (253)
          Payload: STATUSTEXT (253)
              severity (MAV_SEVERITY): MAV_SEVERITY_WARNING (4)
              text (char): Mode change to LOITER failed: requires position
          Message CRC: 0x647a
      
      Frame 3: 101 bytes on wire (808 bits), 101 bytes captured (808 bits) on 
      Null/Loopback
      Internet Protocol Version 4, Src: 127.0.0.1, Dst: 127.0.0.1
      Transmission Control Protocol, Src Port: 62025, Dst Port: 5760, Seq: 706
      MAVLink Protocol (45)
          Header
              Magic value / version: MAVLink 2.0 (0xfd)
              Payload length: 33
              Incompatibility flag: 0
              Compatibility flag: 0
              Packet sequence: 11
              System id: 0xfa
              Component id: 0xfa
              Message id: COMMAND_LONG (76)
          Payload: COMMAND_LONG (76)
              target_system (uint8_t): 1
              target_component (uint8_t): 1
              command (MAV_CMD): MAV_CMD_COMPONENT_ARM_DISARM (400)
              confirmation (uint8_t): 1
              param1: Arm (float): 0
              param2: Force (float): 0
              param3 (float): 0
              param4 (float): 0
              param5 (float): 0
              param6 (float): 0
              param7 (float): 0
          Message CRC: 0xf9df
      
      Frame 4: 144 bytes on wire (1152 bits), 144 bytes captured (1152 bits) o
      Null/Loopback
      Internet Protocol Version 4, Src: 127.0.0.1, Dst: 127.0.0.1
      Transmission Control Protocol, Src Port: 5760, Dst Port: 62025, Seq: 126
      MAVLink Protocol (88)
          Header
              Magic value / version: MAVLink 2.0 (0xfd)
              Payload length: 10
              Incompatibility flag: 0
              Compatibility flag: 0
              Packet sequence: 71
              System id: 0x01
              Component id: 0x01
              Message id: COMMAND_ACK (77)
          Payload: COMMAND_ACK (77)
              command (MAV_CMD): MAV_CMD_COMPONENT_ARM_DISARM (400)
              result (MAV_RESULT): MAV_RESULT_ACCEPTED (0)
          Message CRC: 0x1878
      MAVLink Protocol (88)
          Header
              Magic value / version: MAVLink 2.0 (0xfd)
              Payload length: 17
              Incompatibility flag: 0
              Compatibility flag: 0
              Packet sequence: 72
              System id: 0x01
              Component id: 0x01
              Message id: STATUSTEXT (253)
          Payload: STATUSTEXT (253)
              severity (MAV_SEVERITY): MAV_SEVERITY_INFO (6)
              text (char): Disarming motors
          Message CRC: 0x570a
      MAVLink Protocol (88)
          Header
              Magic value / version: MAVLink 2.0 (0xfd)
              Payload length: 25
              Incompatibility flag: 0
              Compatibility flag: 0
              Packet sequence: 73
              System id: 0x01
              Component id: 0x01
              Message id: PARAM_VALUE (22)
          Payload: PARAM_VALUE (22)
              param_id (char): MOT_THST_HOVER
              param_value (float): 0.39
              param_type (MAV_PARAM_TYPE): MAV_PARAM_TYPE_REAL32 (9)
              param_count (uint16_t): 1305
              param_index (uint16_t): 65535
          Message CRC: 0xf478
      
  
    ''',
  })
# ---
# name: test_wlua[common.xml-common.pcapng]
  dict({
    'returncode': 0,
    'stderr': '',
    'stdout': '''
      Frame 1: 101 bytes on wire (808 bits), 101 bytes captured (808 bits) on 
      Null/Loopback
      Internet Protocol Version 4, Src: 127.0.0.1, Dst: 127.0.0.1
      Transmission Control Protocol, Src Port: 62025, Dst Port: 5760, Seq: 1, 
      MAVLink Protocol (45)
          Header
              Magic value / version: MAVLink 2.0 (0xfd)
              Payload length: 33
              Incompatibility flag: 0
              Compatibility flag: 0
              Packet sequence: 246
              System id: 0xfa
              Component id: 0xfa
              Message id: COMMAND_LONG (76)
          Payload: COMMAND_LONG (76)
              target_system (uint8_t): 1
              target_component (uint8_t): 1
              command (MAV_CMD): MAV_CMD_DO_SET_MODE (176)
              confirmation (uint8_t): 1
              param1: Mode (MAV_MODE): Unknown (1)
              param2: Custom Mode (float): 5
              param3: Custom Submode (float): 0
              param4 (float): 0
              param5 (float): 0
              param6 (float): 0
              param7 (float): 0
          Message CRC: 0xdb93
      
      Frame 2: 138 bytes on wire (1104 bits), 138 bytes captured (1104 bits) o
      Null/Loopback
      Internet Protocol Version 4, Src: 127.0.0.1, Dst: 127.0.0.1
      Transmission Control Protocol, Src Port: 5760, Dst Port: 62025, Seq: 1, 
      MAVLink Protocol (82)
          Header
              Magic value / version: MAVLink 2.0 (0xfd)
              Payload length: 10
              Incompatibility flag: 0
              Compatibility flag: 0
              Packet sequence: 218
              System id: 0x01
              Component id: 0x01
              Message id: COMMAND_ACK (77)
          Payload: COMMAND_ACK (77)
              command (MAV_CMD): MAV_CMD_DO_SET_MODE (176)
              result (MAV_RESULT): MAV_RESULT_FAILED (4)
          Message CRC: 0x7692
      MAVLink Protocol (82)
          Header
              Magic value / version: MAVLink 2.0 (0xfd)
              Payload length: 48
              Incompatibility flag: 0
              Compatibility flag: 0
              Packet sequence: 219
              System id: 0x01
              Component id: 0x01
              Message id: STATUSTEXT (253)
          Payload: STATUSTEXT (253)
              severity (MAV_SEVERITY): MAV_SEVERITY_WARNING (4)
              text (char): Mode change to LOITER failed: requires position
          Message CRC: 0x647a
      
      Frame 3: 101 bytes on wire (808 bits), 101 bytes captured (808 bits) on 
      Null/Loopback
      Internet Protocol Version 4, Src: 127.0.0.1, Dst: 127.0.0.1
      Transmission Control Protocol, Src Port: 62025, Dst Port: 5760, Seq: 706
      MAVLink Protocol (45)
          Header
              Magic value / version: MAVLink 2.0 (0xfd)
              Payload length: 33
              Incompatibility flag: 0
              Compatibility flag: 0
              Packet sequence: 11
              System id: 0xfa
              Component id: 0xfa
              Message id: COMMAND_LONG (76)
          Payload: COMMAND_LONG (76)
              target_system (uint8_t): 1
              target_component (uint8_t): 1
              command (MAV_CMD): MAV_CMD_COMPONENT_ARM_DISARM (400)
              confirmation (uint8_t): 1
              param1: Arm (float): 0
              param2: Force (float): 0
              param3 (float): 0
              param4 (float): 0
              param5 (float): 0
              param6 (float): 0
              param7 (float): 0
          Message CRC: 0xf9df
      
      Frame 4: 144 bytes on wire (1152 bits), 144 bytes captured (1152 bits) o
      Null/Loopback
      Internet Protocol Version 4, Src: 127.0.0.1, Dst: 127.0.0.1
      Transmission Control Protocol, Src Port: 5760, Dst Port: 62025, Seq: 126
      MAVLink Protocol (88)
          Header
              Magic value / version: MAVLink 2.0 (0xfd)
              Payload length: 10
              Incompatibility flag: 0
              Compatibility flag: 0
              Packet sequence: 71
              System id: 0x01
              Component id: 0x01
              Message id: COMMAND_ACK (77)
          Payload: COMMAND_ACK (77)
              command (MAV_CMD): MAV_CMD_COMPONENT_ARM_DISARM (400)
              result (MAV_RESULT): MAV_RESULT_ACCEPTED (0)
          Message CRC: 0x1878
      MAVLink Protocol (88)
          Header
              Magic value / version: MAVLink 2.0 (0xfd)
              Payload length: 17
              Incompatibility flag: 0
              Compatibility flag: 0
              Packet sequence: 72
              System id: 0x01
              Component id: 0x01
              Message id: STATUSTEXT (253)
          Payload: STATUSTEXT (253)
              severity (MAV_SEVERITY): MAV_SEVERITY_INFO (6)
              text (char): Disarming motors
          Message CRC: 0x570a
      MAVLink Protocol (88)
          Header
              Magic value / version: MAVLink 2.0 (0xfd)
              Payload length: 25
              Incompatibility flag: 0
              Compatibility flag: 0
              Packet sequence: 73
              System id: 0x01
              Component id: 0x01
              Message id: PARAM_VALUE (22)
          Payload: PARAM_VALUE (22)
              param_id (char): MOT_THST_HOVER
              param_value (float): 0.39
              param_type (MAV_PARAM_TYPE): MAV_PARAM_TYPE_REAL32 (9)
              param_count (uint16_t): 1305
              param_index (uint16_t): 65535
          Message CRC: 0xf478
      
  
    ''',
  })
# ---
# name: test_wlua[minimal.xml-common.pcapng]
  dict({
    'returncode': 0,
    'stderr': '',
    'stdout': '''
      Frame 1: 101 bytes on wire (808 bits), 101 bytes captured (808 bits) on 
      Null/Loopback
      Internet Protocol Version 4, Src: 127.0.0.1, Dst: 127.0.0.1
      Transmission Control Protocol, Src Port: 62025, Dst Port: 5760, Seq: 1, 
      MAVLink Protocol (45)
          Header
              Magic value / version: MAVLink 2.0 (0xfd)
              Payload length: 33
              Incompatibility flag: 0
              Compatibility flag: 0
              Packet sequence: 246
              System id: 0xfa
              Component id: 0xfa
              Message id: Unknown (76)
          [Expert Info (Error/Malformed): Unknown message type]
              [Unknown message type]
              [Severity level: Error]
              [Group: Malformed]
          Unparsable payload: 0000803f0000a04000000000000000000000000000000000
          Message CRC: 0xdb93
      
      Frame 2: 138 bytes on wire (1104 bits), 138 bytes captured (1104 bits) o
      Null/Loopback
      Internet Protocol Version 4, Src: 127.0.0.1, Dst: 127.0.0.1
      Transmission Control Protocol, Src Port: 5760, Dst Port: 62025, Seq: 1, 
      MAVLink Protocol (82)
          Header
              Magic value / version: MAVLink 2.0 (0xfd)
              Payload length: 10
              Incompatibility flag: 0
              Compatibility flag: 0
              Packet sequence: 218
              System id: 0x01
              Component id: 0x01
              Message id: Unknown (77)
          [Expert Info (Error/Malformed): Unknown message type]
              [Unknown message type]
              [Severity level: Error]
              [Group: Malformed]
          Unparsable payload: b000040000000000fafa9276fd300000db0101fd0000044d
          Message CRC: 0x647a
      
      Frame 3: 101 bytes on wire (808 bits), 101 bytes captured (808 bits) on 
      Null/Loopback
      Internet Protocol Version 4, Src: 127.0.0.1, Dst: 127.0.0.1
      Transmission Control Protocol, Src Port: 62025, Dst Port: 5760, Seq: 706
      MAVLink Protocol (45)
          Header
              Magic value / version: MAVLink 2.0 (0xfd)
              Payload length: 33
              Incompatibility flag: 0
              Compatibility flag: 0
              Packet sequence: 11
              System id: 0xfa
              Component id: 0xfa
              Message id: Unknown (76)
          [Expert Info (Error/Malformed): Unknown message type]
              [Unknown message type]
              [Severity level: Error]
              [Group: Malformed]
          Unparsable payload: 000000000000000000000000000000000000000000000000
          Message CRC: 0xf9df
      
      Frame 4: 144 bytes on wire (1152 bits), 144 bytes captured (1152 bits) o
      Null/Loopback
      Internet Protocol Version 4, Src: 127.0.0.1, Dst: 127.0.0.1
      Transmission Control Protocol, Src Port: 5760, Dst Port: 62025, Seq: 126
      MAVLink Protocol (88)
          Header
              Magic value / version: MAVLink 2.0 (0xfd)
              Payload length: 10
              Incompatibility flag: 0
              Compatibility flag: 0
              Packet sequence: 71
              System id: 0x01
              Component id: 0x01
              Message id: Unknown (77)
          [Expert Info (Error/Malformed): Unknown message type]
              [Unknown message type]
              [Severity level: Error]
              [Group: Malformed]
          Unparsable payload: 9001000000000000fafa7818fd110000480101fd00000644
          Message CRC: 0xf478
      
  
    ''',
  })
# ---
